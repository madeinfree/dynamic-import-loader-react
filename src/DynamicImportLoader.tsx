import * as React from 'react';
import {
  DynamicImportProps,
  DynamicImportState
} from './DynamicImportLoader.d';

class DynamicImportLoader extends React.Component<
  DynamicImportProps,
  DynamicImportState
> {
  constructor(props: DynamicImportProps) {
    super(props);
    this.state = {
      BaseInstance: null
    };
  }
  componentDidMount() {
    const { fileImport }: DynamicImportProps = this.props;
    fileImport().then((file: any) => {
      this.setState(state => ({
        BaseInstance: file.default
      }));
    });
  }
  render() {
    const { BaseInstance } = this.state;
    const { jsxElement, renderElement: RenderElement } = this.props;
    const children = this.props.children;
    if (BaseInstance) {
      if (children) {
        return children(BaseInstance);
      } else {
        return <BaseInstance />;
      }
    } else {
      if (jsxElement) {
        return jsxElement;
      } else if (RenderElement) {
        return <RenderElement />;
      }
    }
    return <div>Loading...</div>;
  }
}

export default DynamicImportLoader;
