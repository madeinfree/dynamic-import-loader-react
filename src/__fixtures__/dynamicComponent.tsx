import * as React from 'react';

type Props = {};
const DynamicComponent: React.StatelessComponent = (props: Props) => (
  <div>Hello World</div>
);

export default DynamicComponent;
