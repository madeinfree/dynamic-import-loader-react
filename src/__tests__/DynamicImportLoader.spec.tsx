import * as React from 'react';
import DynamicImportLoader from '../DynamicImportLoader';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { mount, shallow, ShallowWrapper, ReactWrapper } from 'enzyme';
import * as renderer from 'react-test-renderer';
import * as jsdom from 'jsdom';
import dynamicComponent from '../__fixtures__/dynamicComponent';

declare var global: NodeJS.Global & {
  document: Document;
};

const doc: Document = new jsdom.JSDOM(
  '<!doctype html><html><body></body></html>'
).window.document;
global.document = doc;

Enzyme.configure({ adapter: new Adapter() });

describe('Dynamic Import Loader', () => {
  it('dynamicImportLoader snapshot', () => {
    const dynamicComponent = renderer.create(
      <DynamicImportLoader
        jsxElement={<div>Loading...</div>}
        fileImport={() => import('../__fixtures__/dynamicComponent')}
      />
    );
    const tree = dynamicComponent.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('dynamicImportLoader loading throw error without loading props', () => {
    try {
      shallow(
        <DynamicImportLoader
          fileImport={() => import('../__fixtures__/dynamicComponent')}
        />
      );
    } catch (err) {
      expect(err.message).toEqual(
        '[Dynamic Import Loder] should specified loading component when useLoading is not true.'
      );
    }
  });

  it('dynamicImportLoader props with jsxElement', done => {
    const Hello: JSX.Element = <div>Custom Loader</div>;
    const tree: ReactWrapper = mount(
      <DynamicImportLoader
        jsxElement={Hello}
        fileImport={() => import('../__fixtures__/dynamicComponent')}
      />
    );
    expect(tree.find('div').text()).toEqual('Custom Loader');

    process.nextTick(() => {
      tree.update();
      expect(tree.find(dynamicComponent).text()).toEqual('Hello World');
      done();
    });
  });

  it('dynamicImportLoader props with renderElement ', done => {
    const tree: ReactWrapper = mount(
      <DynamicImportLoader
        fileImport={() => import('../__fixtures__/dynamicComponent')}
        renderElement={() => <p>Yo</p>}
      />
    );
    expect(tree.find('p').text()).toEqual('Yo');

    process.nextTick(() => {
      tree.update();
      expect(tree.find(dynamicComponent).text()).toEqual('Hello World');
      done();
    });
  });

  it('Dynamic Import Loader render function 使用', done => {
    let tree: ReactWrapper = mount(
      <DynamicImportLoader
        jsxElement={<div>Loading</div>}
        fileImport={() => import('../__fixtures__/dynamicComponent')}
      >
        {Component => (
          <div>
            <button>Click Me</button>
            <Component />
          </div>
        )}
      </DynamicImportLoader>
    );

    process.nextTick(() => {
      tree.update();
      expect(tree.find('button').text()).toEqual('Click Me');
      done();
    });
  });
});
