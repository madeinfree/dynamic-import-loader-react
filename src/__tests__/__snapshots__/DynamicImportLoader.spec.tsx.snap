// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dynamic Import Loader dynamicImportLoader snapshot 1`] = `
<div>
  Loading...
</div>
`;
